cpp/basix/finite-element.cpp

```diff
@@ -322,6 +322,189 @@ basix::create_element(element::family, cell::type, int,
+{
+  std::vector<int> dof_ordering = tp_dof_ordering(
+      family, cell, degree, lvariant, dvariant, discontinuous);
+  return create_element<T>(family, cell, degree, lvariant, dvariant,
+                           discontinuous, dof_ordering);
+}
+//-----------------------------------------------------------------------------
+template basix::FiniteElement<float>
+basix::create_tp_element(element::family, cell::type, int,
+                         element::lagrange_variant, element::dpc_variant, bool);
+template basix::FiniteElement<double>
+basix::create_tp_element(element::family, cell::type, int,
+                         element::lagrange_variant, element::dpc_variant, bool);
+//-----------------------------------------------------------------------------
+template <std::floating_point T>
+std::vector<std::vector<FiniteElement<T>>>
+basix::tp_factors(element::family family, cell::type cell, int degree,
+                  element::lagrange_variant lvariant,
+                  element::dpc_variant dvariant, bool discontinuous,
+                  std::vector<int> dof_ordering)
+{
+  std::vector<int> tp_dofs = tp_dof_ordering(family, cell, degree, lvariant,
+                                             dvariant, discontinuous);
+  if (tp_dofs.size() > 0 && tp_dofs == dof_ordering)
+  {
+    switch (family)
+    {
+    case element::family::P:
+    {
+      FiniteElement<T> sub_element
+          = create_element<T>(element::family::P, cell::type::interval, degree,
+                              lvariant, dvariant, true);
+      switch (cell)
+      {
+      case cell::type::quadrilateral:
+      {
+        return {{sub_element, sub_element}};
+      }
+      case cell::type::hexahedron:
+      {
+        return {{sub_element, sub_element, sub_element}};
+      }
+      default:
+      {
+        throw std::runtime_error("Invalid celltype.");
+      }
+      }
+      break;
+    }
+    default:
+    {
+      throw std::runtime_error("Invalid family.");
+    }
+    }
+  }
+  return {};
+}
+//-----------------------------------------------------------------------------
+template std::vector<std::vector<basix::FiniteElement<float>>>
+basix::tp_factors(element::family, cell::type, int, element::lagrange_variant,
+                  element::dpc_variant, bool, std::vector<int>);
+template std::vector<std::vector<basix::FiniteElement<double>>>
+basix::tp_factors(element::family, cell::type, int, element::lagrange_variant,
+                  element::dpc_variant, bool, std::vector<int>);
+//-----------------------------------------------------------------------------
+std::vector<int> basix::tp_dof_ordering(element::family family, cell::type cell,
+                                        int degree, element::lagrange_variant,
+                                        element::dpc_variant, bool)
+{
+  std::vector<int> dof_ordering;
+  std::vector<int> perm;
+
+  switch (family)
+  {
+  case element::family::P:
```

Comment:
Should this throw an error for other families? Currently just returning `{}` I think?

Reply:
Yes, throwing is probably better than returning an empty list